go-controller/pkg/clustermanager/userdefinednetwork/controller.go

```diff
@@ -112,19 +118,112 @@ func (c *Controller) reconcile(key string) error {
 	udnCopy := udn.DeepCopy()
 	nadCopy := nad.DeepCopy()
 
-	syncErr := c.syncUserDefinedNetwork(udnCopy, nadCopy)
+	nadCopy, syncErr := c.syncUserDefinedNetwork(udnCopy, nadCopy)
 
 	updateStatusErr := c.updateUserDefinedNetworkStatus(udnCopy, nadCopy, syncErr)
 
 	return errors.Join(syncErr, updateStatusErr)
 }
 
-func (c *Controller) syncUserDefinedNetwork(udn *userdefinednetworkv1.UserDefinedNetwork, nad *netv1.NetworkAttachmentDefinition) error {
-	//TODO: implement sync UDN objects logic
-	return errors.New("implement me")
+func (c *Controller) syncUserDefinedNetwork(udn *userdefinednetworkv1.UserDefinedNetwork, nad *netv1.NetworkAttachmentDefinition) (*netv1.NetworkAttachmentDefinition, error) {
+	if udn == nil {
+		return nil, nil
+	}
+
+	desiredNAD, err := c.renderNadFn(udn)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate NetworkAttachmentDefinition: %w", err)
+	}
+	if nad == nil {
+		nad, err = c.nadClient.K8sCniCncfIoV1().NetworkAttachmentDefinitions(udn.Namespace).Create(context.Background(), desiredNAD, metav1.CreateOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("failed to create NetworkAttachmentDefinition: %w", err)
+		}
+		klog.Infof("Created NetworkAttachmentDefinition [%s/%s]", nad.Namespace, nad.Name)
+		return nad, nil
+	}
+
+	if !metav1.IsControlledBy(nad, udn) {
+		return nil, fmt.Errorf("foreign NetworkAttachmentDefinition with the desired name already exist [%s/%s]", nad.Namespace, nad.Name)
+	}
+
+	if !reflect.DeepEqual(nad.Spec, desiredNAD.Spec) ||
```

Comment:
Originally it worked like that, following previous review I was asked to changed the behavior in a way that if the managed NAD spec / onwer-refrences are mutated, it should be reconciled by the controller, please see https://github.com/ovn-org/ovn-kubernetes/pull/4486#discussion_r1679206695 
In addition having the controller reconcile the NAD it creates prevent incomplete state where NAD spec is gen2 but pod created using NAD from gen 1, in case the NAD spec is changed manually.
WDYT?

Reply:
+1 to raising an error and not doing anything.

this way, we're sure we're not handing a user what they don't want.

Reply:
In a scenario UDN is created, the controller created a NAD, now the NAD is muated from some reason by on user or another, the NAD is no longer match the desired spec.
The controller should reconcile the NAD and mutate it to fit the desired state, similar to any controller that manage resources, applying to kubernetes controllers convention.

Nevertheless, there is also the scenerio I mentioned that should be avoided:
```
In addition having the controller reconcile the NAD it creates prevent incomplete state where NAD spec is gen2 but pod created using NAD from gen 1, in case the NAD spec is changed manually.
```

Reply:
> In a scenario UDN is created, the controller created a NAD, now the NAD is muated from some reason by on user or another, the NAD is no longer match the desired spec. The controller should reconcile the NAD and mutate it to fit the desired state, similar to any controller that manage resources, applying to kubernetes controllers convention.

Well, we're not speaking about *a* user. It's an admin. 

And - IIUC - in this particular scenario, your controller is taking over an **existing** NAD. NAD updates are **not** allowed. Thus, the way I see it, your controller is mutating an existing NAD. Which is not allowed. 

Am I missing something ? 

> 
> Nevertheless, there is also the scenerio I mentioned that should be avoided:
> 
> ```
> In addition having the controller reconcile the NAD it creates prevent incomplete state where NAD spec is gen2 but pod created using NAD from gen 1, in case the NAD spec is changed manually.
> ```

I don't follow this sentence, but, NAD updates are not allowed. OVN-K flat out rejects them, going as far as breaking all the controllers in the network.

Reply:
So @npinaeva points out that several lines above this we have 
```
	if !metav1.IsControlledBy(nad, udn) {
		return nil, fmt.Errorf("foreign NetworkAttachmentDefinition with the desired name already exist [%s/%s]", nad.Namespace, nad.Name)
	}
```

So a nad that is not owned by the udn cannot be modified. That resolves that issue. Additionally she points out with CEL we do not allow updates to the CRD:

https://github.com/ovn-org/ovn-kubernetes/pull/4486/files#diff-325e05028094fe7cf6da81ead819b9137533b857a1fb1d54dc33752c03ff1804R35

That leaves one last thing, which is I don't get why we have this condition in the if statement:
```
!reflect.DeepEqual(nad.ObjectMeta.OwnerReferences, desiredNAD.ObjectMeta.OwnerReferences)
```

We already checked that the nad has the same owner/controller as UDN before we got to this code. But now we want to make sure all the owner references are exactly the same, why?

Reply:
> Am I missing something ?

The controller does not take over existing NAD, it creates UDN NADs and managed them only.
Managed NAD is one that created by the controller and has matching owner-reference and finalizer.

> I don't follow this sentence, but, NAD updates are not allowed. OVN-K flat out rejects them, going as far as breaking all the controllers in the network.

This is exactly why this logic is crucial, reconciling the managed NAD according to spec, prevent managed NAD from being mutated.

Reply:
> That leaves one last thing, which is I don't get why we have this condition in the if statement:
> !reflect.DeepEqual(nad.ObjectMeta.OwnerReferences, desiredNAD.ObjectMeta.OwnerReferences)
> We already checked that the nad has the same owner/controller as UDN before we got to this code. But now we want to make sure all the owner references are exactly the same, why

I think this condition is necessary to ensure a managed NAD has the expected owner-reference, similar to the spec nothing prevents from one to mutate the owner-references.
It prevent the scenario the owner-reference is removed or changes in a way the controller does not identify the NAD as a managed one.

Reply:
but in that case we would have failed the check several lines above that checks if the nad has an owner reference to UDN and it is the controller. Then we would have already returned, so I don't see the value in this check.

Reply:
Make sense, thanks for explaining, I will drop this check.